File: src/api/python/api/fog05/interfaces/RuntimePluginFDU.py
# Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
# which is available at https://www.apache.org/licenses/LICENSE-2.0.
#
# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
#
# Contributors: Gabriele Baldoni, ADLINK Technology Inc.
# OCaml implementation and API


import uuid
import time
from fog05.interfaces.Plugin import Plugin
from fog05 import Yaks_Connector
from fog05 import Yaks_Connector
from fog05.DLogger import DLogger
from fog05.interfaces.InfraFDU import InfraFDU

class RuntimePluginFDU(Plugin):

    def __init__(self,name, version, plugin_uuid, manifest):
        super(RuntimePluginFDU, self).__init__(version, plugin_uuid)
        self.pid = -1
        self.name = name
        loc = manifest.get('configuration').get('ylocator').split('/')[1]
        self.connector = Yaks_Connector(loc)
        self.logger = DLogger(debug_flag=True)
        self.node = manifest.get('configuration').get('nodeid')
        self.manifest = manifest
        self.configuration = manifest.get('configuration',{})

        self.current_fdus = {}

    def wait_destination_ready(self, fduid, instanceid, destinationid):
        flag = False
        while not flag:
            try:
                res = self.agent.get_node_fdu_info(fduid, instanceid, destinationid)
                while res.get('status') != 'LAND':
                    time.sleep(0.250)
                    res = self.agent.get_node_fdu_info(fduid, instanceid, destinationid)
                flag = True
            except:
                pass
        return flag


    def wait_dependencies(self):
        self.get_agent()
        os = None
        while os is None:
            try:
                os = self.get_os_plugin()
            except (RuntimeError, ValueError):
                time.sleep(1)
        nm = None
        while nm is None:
            try:
                nm = self.get_nm_plugin()
            except (RuntimeError, ValueError):
                time.sleep(1)
        return


    def write_fdu_error(self, fdu_uuid, instance_uuid, errno, errmsg):
        record = self.connector.loc.actual.get_node_fdu(self.node, self.uuid, fdu_uuid, instance_uuid)
        if record is None:
            record = self.connector.loc.desired.get_node_fdu(self.node, self.uuid, fdu_uuid, instance_uuid)
        record = InfraFDU(record)
        record.set_status('ERROR')
        record.set_error_code(errno)
        record.set_error_msg('{}'.format(errmsg))
        self.connector.loc.actual.add_node_fdu(self.node, self.uuid, fdu_uuid, instance_uuid, record.to_json())

    def update_fdu_status(self, fdu_uuid, instance_uuid, status):
        record = self.connector.loc.actual.get_node_fdu(self.node, self.uuid, fdu_uuid, instance_uuid)
        if record is None:
            record = self.connector.loc.desired.get_node_fdu(self.node, self.uuid, fdu_uuid, instance_uuid)
        record = InfraFDU(record)
        record.set_status(status)
        self.connector.loc.actual.add_node_fdu(self.node, self.uuid, fdu_uuid,instance_uuid, record.to_json())

    def get_local_instances(self, fdu_uuid):
        return self.connector.loc.actual.get_node_fdu_instances(self.node, fdu_uuid)

    def start_runtime(self):
        '''
        start the runtime
        :return: runtime pid or runtime uuid?
        '''
        raise NotImplementedError('This is and interface!')

    def stop_runtime(self):
        '''
        stop this runtime
        '''
        raise NotImplementedError('This is and interface!')

    def get_fdus(self):
        raise NotImplementedError('This is and interface!')

    def define_fdu(self, fdu_manifest):
        '''
        Define fdu from args of from manifest file passed within parameters
        return the fdu uuid
        :args: manifest of the FDU

        ##### ASSUMING FDU MANIFEST = old ATOMIC ENTITY MANIFEST

        :return: String
        '''

        raise NotImplementedError('This is and interface!')

    def undefine_fdu(self, fdu_uuid):
        '''
        Undefine the fdu identified by fdu_uuid
        if the fdu state do not allow transition to UNDEFINED
        should throw an exception

        :fdu_uuid: String
        :return: bool

        '''
        raise NotImplementedError('This is and interface!')

    def run_fdu(self, fdu_uuid):
        '''
        Start the fdu identified by fdu_uuid
        if the fdu state do not allow transition to RUN (eg is non CONFIGURED)
        should throw an exception

        :fdu_uuid: String
        :return: bool

        '''
        raise NotImplementedError('This is and interface!')

    def stop_fdu(self, fdu_uuid):
        '''
        Stop the fdu identified by fdu_uuid
        if the fdu state do not allow transition to CONFIGURED
        (in this case is not in RUNNING)
        should throw an exception

        :fdu_uuid: String
        :return: bool

        '''
        raise NotImplementedError('This is and interface!')

    def migrate_fdu(self, fdu_uuid, dst=False):
        '''
        Migrate the fdu identified by fdu_uuid to the new FogNode
        identified by fognode_uuid
        The migration depend to the nature of the fdu
         (native app, µSvc, VM, Container)
        if the fdu state do not allow transition to
         MIGRATE (eg a native app can't be migrated)
        should throw an exception
        if the destination node can't handle the
         migration an exception should be throwed

        To help migration should use the two methods:

        - before_migrate_fdu_actions()
        - after_migrate_fdu_actions()

        After the migration the fdu on the source node has to be undefined
        the one on the destination node has to be in RUNNING


        :fdu_uuid: String
        :fognode_uuid: String
        :return: bool

        '''

        raise NotImplementedError('This is and interface!')

    def before_migrate_fdu_actions(self, fdu_uuid, dst=False):
        '''
        Action to be taken before a migration
        eg. copy disks of vms, save state of µSvc

        :fdu_uuid: String
        :return: bool

        '''

        raise NotImplementedError('This is and interface!')

    def after_migrate_fdu_actions(self, fdu_uuid, dst=False):
        '''
        Action to be taken after a migration
        eg. delete disks of vms, delete state of µSvc, undefine fdu

        :fdu_uuid: String
        :return: bool

        '''

        raise NotImplementedError('This is and interface!')

    def scale_fdu(self, fdu_uuid):
        '''
        Scale an fdu
        eg. give more cpu/ram/disk, maybe passed
         by parameter the new scale value?

        if the fdu state do not allow transition to
        SCALE (in this case is not in RUNNING)
        should throw an exception

        :fdu_uuid: String
        :return: bool


        '''
        raise NotImplementedError('This is and interface!')

    def pause_fdu(self, fdu_uuid):
        '''
        Pause an fdu

        if the fdu state do not allow transition
        to PAUSED (in this case is not in RUNNING)
        should throw an exception

        :fdu_uuid: String
        :return: bool


        '''
        raise NotImplementedError('This is and interface!')

    def resume_fdu(self, fdu_uuid):
        '''
        Resume an fdu

        if the fdu state do not allow transition to
         RUNNING (in this case is not in PAUSED)
        should throw an exception

        :fdu_uuid: String
        :return: bool


        '''
        raise NotImplementedError('This is and interface!')

    def configure_fdu(self, fdu_uuid):
        '''
        TBD

        :fdu_uuid: String
        :return: bool


        '''
        raise NotImplementedError('This is and interface!')

    def clean_fdu(self, fdu_uuid):
        '''
        Clean an fdu

        this mode destroy the fdu instance


        if the fdu state do not allow transition
        to DEFINED (in this case is not in CONFIGURED)
        should throw an exception

        :fdu_uuid: String
        :return: bool


        '''
        raise NotImplementedError('This is and interface!')

    def is_uuid(self, uuid_string):
        try:
            val = uuid.UUID(uuid_string, version=4)
        except ValueError:
            return False
        return True


class FDUNotExistingException(Exception):
    def __init__(self, message, errors=0):

        super(FDUNotExistingException, self).__init__(message)
        self.errors = errors


class MigrationNotPossibleException(Exception):
    def __init__(self, message, errors=0):

        super(MigrationNotPossibleException, self).__init__(message)
        self.errors = errors


class MigrationNotAllowedException(Exception):
    def __init__(self, message, errors=0):

        super(MigrationNotAllowedException, self).__init__(message)
        self.errors = errors


class StateTransitionNotAllowedException(Exception):
    def __init__(self, message, errors=0):

        super(StateTransitionNotAllowedException, self).__init__(message)
        self.errors = errors

File: src/api/go/fog05-go/runtime_plugin_fdu.go
package fog05

import (
	"fmt"
	"time"

	"github.com/google/uuid"
	log "github.com/sirupsen/logrus"
)

// FOSRuntimePluginInterface is the interface to be implenter for a Runtime Plugin
type FOSRuntimePluginInterface interface {

	//StartRuntime starts the plugin
	StartRuntime() error

	//StopRuntime stops the plugin
	StopRuntime() error

	//GetFDUs gets all FDU and instances information
	GetFDUs() map[string]FDURecord

	//DefineFDU defines an FDU instance from the given record
	DefineFDU(FDURecord) error

	//UndefineFDU undefines the given FDU instance
	UndefineFDU(string) error

	//ConfigureFDU configures the given FDU instance
	ConfigureFDU(string) error

	//CleanFDU cleans the given FDU instance
	CleanFDU(string) error

	//RunFDU starts the given FDU instance
	RunFDU(string) error

	//StopFDU stops the given FDU instance
	StopFDU(string) error

	//MigrateFDU migrates the given FDU instance
	MigrateFDU(string) error

	//ScaleFDU scales the given FDU instance
	ScaleFDU(string) error

	//PauseFDU pauses the given FDU instance
	PauseFDU(string) error

	//ResumeFDU resumes the given FDU instance
	ResumeFDU(string) error
}

// FOSRuntimePluginAbstract represents a Runtime Plugin for Eclipse fog05
type FOSRuntimePluginAbstract struct {
	Pid           int
	Name          string
	Connector     *YaksConnector
	Node          string
	Configuration map[string]string
	// Plugin        *FOSPlugin
	Logger *log.Logger
	FOSRuntimePluginInterface
	FOSPlugin
}

// NewFOSRuntimePluginAbstract returns a new FOSRuntimePluginFDU object
func NewFOSRuntimePluginAbstract(name string, version int, pluginid string, manifest Plugin) (*FOSRuntimePluginAbstract, error) {
	if pluginid == "" {
		pluginid = uuid.UUID.String(uuid.New())
	}
	pl := NewPlugin(version, pluginid)

	conf := *manifest.Configuration
	// json.Unmarshal([]byte(manifest.Configuration), &conf)
	con, err := NewYaksConnector(conf["ylocator"].(string))
	if err != nil {
		return nil, err
	}
	pl.connector = con
	pl.node = conf["nodeid"].(string)

	return &FOSRuntimePluginAbstract{Pid: -1, Name: name, Connector: con, Node: conf["nodeid"].(string), FOSPlugin: *pl, Logger: log.New()}, nil
}

// Start starts the Plugin and calls StartRuntime of FOSRuntimePluginInterface
func (rt *FOSRuntimePluginAbstract) Start() {
	rt.WaitDependencies()
	rt.Connector.Local.Desired.ObserveNodeRuntimeFDU(rt.Node, rt.FOSPlugin.UUID, rt.react)
	err := rt.FOSRuntimePluginInterface.StartRuntime()
	if err != nil {
		rt.Logger.Error(fmt.Sprintf("Plugin StartRuntime returned error %s", err.Error()))
		rt.Close()
	}
}

// Close closes the Plugin, called by FOSRuntimePluginInterface.StopRuntime()
func (rt *FOSRuntimePluginAbstract) Close() {
	rt.RemovePlugin()
	rt.Connector.Close()
	rt.Logger.Info("Plugin closed")
}

// WaitDestinationReady waits for the destination node of a migration to be ready
func (rt *FOSRuntimePluginAbstract) WaitDestinationReady(fduid string, instanceid string, destinationid string) bool {
	return false
}

// WaitDependencies waits that the Agent, OS and NM Plugins are up and gets those from YAKS
func (rt *FOSRuntimePluginAbstract) WaitDependencies() {
	rt.FOSPlugin.GetAgent()
	for rt.FOSPlugin.OS == nil {
		rt.FOSPlugin.GetOSPlugin()
		time.Sleep(1 * time.Second)
	}
	for rt.FOSPlugin.NM == nil {
		rt.FOSPlugin.GetNMPlugin()
		time.Sleep(1 * time.Second)
	}

}

// WriteFDUError given an fdu id, instance id, error number and error message, stores the error in YAKS
func (rt *FOSRuntimePluginAbstract) WriteFDUError(fduid string, instanceid string, errno int, errmsg string) error {
	record, err := rt.Connector.Local.Actual.GetNodeFDU(rt.Node, rt.FOSPlugin.UUID, fduid, instanceid)
	if err != nil {
		return err
	}

	record.Status = ERROR
	record.ErrorCode = &errno
	record.ErrorMsg = &errmsg

	err = rt.Connector.Local.Actual.AddNodeFDU(rt.Node, rt.FOSPlugin.UUID, fduid, instanceid, *record)
	return err
}

// UpdateFDUStatus given an fdu id, instance id and status updates the status in YAKS
func (rt *FOSRuntimePluginAbstract) UpdateFDUStatus(fduid string, instanceid string, status string) error {
	record, err := rt.Connector.Local.Actual.GetNodeFDU(rt.Node, rt.FOSPlugin.UUID, fduid, instanceid)
	if err != nil {
		return err
	}

	record.Status = status

	err = rt.Connector.Local.Actual.AddNodeFDU(rt.Node, rt.FOSPlugin.UUID, fduid, instanceid, *record)
	return err
}

// GetLocalInstances given an fdu id returns all the instances of that fdu running locally, returns string slice
func (rt *FOSRuntimePluginAbstract) GetLocalInstances(fduid string) ([]string, error) {
	return rt.Connector.Local.Actual.GetNodeFDUInstances(rt.Node, fduid)

}

// RegisterPlugin registers the plugin in the node
func (rt *FOSRuntimePluginAbstract) RegisterPlugin(manifest *Plugin) {
	rt.Connector.Local.Actual.AddNodePlugin(rt.Node, rt.FOSPlugin.UUID, *manifest)
}

// RemovePlugin removes the plugin in the node
func (rt *FOSRuntimePluginAbstract) RemovePlugin() {
	rt.Connector.Local.Actual.RemoveNodePlugin(rt.Node, rt.FOSPlugin.UUID)
}

// GetFDUDescriptor retrives an FDURecord for the plugin
func (rt *FOSRuntimePluginAbstract) GetFDUDescriptor(fduid string, instanceid string) (*FDU, error) {
	return rt.FOSPlugin.Agent.GetFDUInfo(rt.Node, fduid, instanceid)
}

// GetFDURecord retrives an FDURecord for the plugin
func (rt *FOSRuntimePluginAbstract) GetFDURecord(instanceid string) (*FDURecord, error) {
	return rt.Connector.Local.Actual.GetNodeFDU(rt.Node, rt.FOSPlugin.UUID, "*", instanceid)
}

// AddFDURecord adds an FDU record to the node
func (rt *FOSRuntimePluginAbstract) AddFDURecord(instanceid string, info *FDURecord) error {
	record, err := rt.Connector.Local.Actual.GetNodeFDU(rt.Node, rt.FOSPlugin.UUID, "*", instanceid)
	if err != nil {
		return err
	}
	return rt.Connector.Local.Actual.AddNodeFDU(rt.Node, rt.FOSPlugin.UUID, record.FDUID, instanceid, *info)
}

// RemoveFDURecord removes an FDURecord from the node
func (rt *FOSRuntimePluginAbstract) RemoveFDURecord(instanceid string) error {
	record, err := rt.Connector.Local.Actual.GetNodeFDU(rt.Node, rt.FOSPlugin.UUID, "*", instanceid)
	if err != nil {
		return err
	}
	return rt.Connector.Local.Actual.RemoveNodeFDU(rt.Node, rt.FOSPlugin.UUID, record.FDUID, instanceid)
}

func (rt *FOSRuntimePluginAbstract) react(info FDURecord) {

	action := info.Status
	id := info.UUID
	switch action {
	case DEFINE:
		rt.DefineFDU(info)
	case UNDEFINE:
		rt.UndefineFDU(id)
	case CLEAN:
		rt.CleanFDU(id)
	case CONFIGURE:
		rt.ConfigureFDU(id)
	case RUN:
		rt.RunFDU(id)
	case STOP:
		rt.StopFDU(id)
	case PAUSE:
		rt.PauseFDU(id)
	case RESUME:
		rt.ResumeFDU(id)
	case LAND:
		rt.MigrateFDU(id)
	case TAKEOFF:
		rt.MigrateFDU(id)
	default:
		rt.Logger.Error(fmt.Sprintf("Action %s not recognized", action))
	}
}
